File: java/src/main/java/org/rocksdb/ByteArrayDeque.java
// Copyright (c) 2014, Facebook, Inc.  All rights reserved.
// This source code is licensed under the BSD-style license found in the
// LICENSE file in the root directory of this source tree. An additional grant
// of patent rights can be found in the PATENTS file in the same directory.

package org.rocksdb;

import java.util.Deque;
import java.util.Collection;
import java.util.Iterator;

/**
 * A Java deque, with elements of type byte[], backed by a JNI c++ deque.
 * 
 * 
 * @see org.rocksdb.AbstractDeque
 */
public abstract class ByteArrayDeque extends AbstractDeque<byte[]> {
  
  @Override public void addFirst(byte[] e) {
    assert(isInitialized());
    addFirst0(nativeHandle_, e);
  }

  @Override public void addLast(byte[] e) {
    assert(isInitialized());
    addLast0(nativeHandle_, e);
  }

  @Override public byte[] removeFirst() {
    assert(isInitialized());
    return removeFirst0(nativeHandle_);
  }

  @Override public byte[] removeLast() {
    assert(isInitialized());
    return removeLast0(nativeHandle_);
  }

  @Override public byte[] pollFirst() {
    assert(isInitialized());
    return pollFirst0(nativeHandle_);
  }

  @Override public byte[] pollLast() {
    assert(isInitialized());
    return pollLast0(nativeHandle_);
  }

  @Override public byte[] getFirst() {
    assert(isInitialized());
    return getFirst0(nativeHandle_);
  }

  @Override public byte[] getLast() {
    assert(isInitialized());
    return getLast0(nativeHandle_);
  }

  @Override public byte[] peekFirst() {
    assert(isInitialized());
    return peekFirst0(nativeHandle_);
  }

  @Override public byte[] peekLast() {
    assert(isInitialized());
    return peekLast0(nativeHandle_);
  }

  @Override public boolean removeFirstOccurrence(Object o) {
    assert(isInitialized());
    return removeFirstOccurrence0(nativeHandle_, o);
  }

  @Override public boolean removeLastOccurrence(Object o) {
    assert(isInitialized());
    return removeLastOccurrence0(nativeHandle_, o);
  }

  @Override public boolean remove(Object o) {
    assert(isInitialized());
    return remove0(nativeHandle_, o);
  }

  @Override public boolean containsAll(Collection<?> c) {
    assert(isInitialized());
    return containsAll0(nativeHandle_, c);
  }

  @Override public boolean addAll(Collection<? extends byte[]> c) {
    assert(isInitialized());
    return addAll0(nativeHandle_, c);
  }

  @Override public boolean removeAll(Collection<?> c) {
    assert(isInitialized());
    return removeAll0(nativeHandle_, c);
  }

  @Override public boolean retainAll(Collection<?> c) {
    assert(isInitialized());
    return retainAll0(nativeHandle_, c);
  }
  
  @Override public void clear() {
    assert (isInitialized());
    clear0(nativeHandle_);
  }

  @Override public boolean contains(Object o) {
    assert(isInitialized());
    return contains0(nativeHandle_, o);
  }
  
  @Override public int size() {
    assert (isInitialized());
    return size0(nativeHandle_);
  }

  @Override public boolean isEmpty() {
    assert (isInitialized());
    return isEmpty0(nativeHandle_);
  }
  
  @Override public abstract Iterator<byte[]> iterator();

  @Override public Object[] toArray() {
    assert(isInitialized());
    return toArray0(nativeHandle_);
  }

  @Override public <T> T[] toArray(T[] a) {
    int size = size();
    if (a.length < size) {
       a = (T[])java.lang.reflect.Array.newInstance(
           a.getClass().getComponentType(), size);
    }
    
    // a has to be an array with "byte[]" elements
    toArray1(nativeHandle_, (Object[]) a);
    
    if (a.length > size) {
      a[size] = null;
    }
    
    return a;
  }

  @Override public abstract Iterator<byte[]> descendingIterator();
  
  /**
   * Creates a string representation of the data
   *
   * @param hex When true, the representation
   *   will be encoded in hexadecimal.
   *
   * @return The string representation of the data.
   */
  public String toString(final boolean hex) {
    assert (isInitialized());
    return toString0(nativeHandle_, hex);
  }

  /**
   * Deletes underlying C++ slice pointer.
   * Note that this function should be called only after all
   * RocksDB instances referencing the slice are closed.
   * Otherwise an undefined behavior will occur.
   */
  @Override
  protected void disposeInternal() {
    assert(isInitialized());
    disposeInternal(nativeHandle_);
  }
  
  private native void addFirst0(long handle, byte[] e);
  private native void addLast0(long handle, byte[] e);
  private native byte[] removeFirst0(long handle);
  private native byte[] removeLast0(long handle);
  private native byte[] pollFirst0(long handle);
  private native byte[] pollLast0(long handle);
  private native byte[] getFirst0(long handle);
  private native byte[] getLast0(long handle);
  private native byte[] peekFirst0(long handle);
  private native byte[] peekLast0(long handle);
  private native boolean removeFirstOccurrence0(long handle, Object o);
  private native boolean removeLastOccurrence0(long handle, Object o);
  private native boolean remove0(long handle, Object o);
  private native boolean containsAll0(long handle, Collection<?> c);
  private native boolean addAll0(long handle, Collection<? extends byte[]> c);
  private native boolean removeAll0(long handle, Collection<?> c);
  private native boolean retainAll0(long handle, Collection<?> c);
  private native void clear0(long handle);
  private native boolean contains0(long handle, Object o);
  private native int size0(long handle);
  private native boolean isEmpty0(long handle);
  
  /////////////////   iter goes here
  
  private native Object[] toArray0(long handle);
  private native Object[] toArray1(long handle, Object[] a);
  
  /////////////////  desc iter goes here
  
  private native String toString0(long handle, boolean hex);

  private native void disposeInternal(long handle);
}

File: java/rocksjni/bytearray_deque.cc
// Copyright (c) 2014, Facebook, Inc.  All rights reserved.
// This source code is licensed under the BSD-style license found in the
// LICENSE file in the root directory of this source tree. An additional grant
// of patent rights can be found in the PATENTS file in the same directory.
//
// This file implements the "bridge" between Java and C++ for
// std::.

#include <jni.h>

#include <deque>
#include <memory>
#include <string>

#include "include/org_rocksdb_ByteArrayDeque.h"
#include "include/org_rocksdb_ByteArrayDeque_Iter.h"
#include "portal.h"

/*
 * Class:     org_rocksdb_ByteArrayDeque
 * Method:    addFirst0
 * Signature: (J[B)V
 */
void JNICALL Java_org_rocksdb_ByteArrayDeque_addFirst0(JNIEnv* env,
                                                       jobject jobj,
                                                       jlong handle,
                                                       jbyteArray elem) {
  // Convert byte[] to std::string
  const auto ptr = rocksdb::JniUtil::byteArrayToStdString(env, elem);
  const auto str = std::shared_ptr<std::string>(ptr);

  // Add to the front of deque
  const auto deque = reinterpret_cast<std::deque<std::string> *>(handle);
  deque->push_front(*str);
}

/*
 * Class:     org_rocksdb_ByteArrayDeque
 * Method:    addLast0
 * Signature: (J[B)V
 */
void JNICALL Java_org_rocksdb_ByteArrayDeque_addLast0(JNIEnv* env, jobject jobj,
                                                      jlong handle,
                                                      jbyteArray elem) {
  // Convert byte[] to std::string
  const auto ptr = rocksdb::JniUtil::byteArrayToStdString(env, elem);
  const auto str = std::shared_ptr<std::string>(ptr);

  // Add to the end of deque
  const auto deque = reinterpret_cast<std::deque<std::string> *>(handle);
  deque->push_back(*str);
}

/*
 * Class:     org_rocksdb_ByteArrayDeque
 * Method:    removeFirst0
 * Signature: (J)[B
 */
jbyteArray JNICALL Java_org_rocksdb_ByteArrayDeque_removeFirst0(JNIEnv* env,
                                                                jobject jobj,
                                                                jlong handle) {
  const auto deque = reinterpret_cast<std::deque<std::string> *>(handle);

  if (deque->empty()) {
    rocksdb::ExceptionJni::ThrowNew(
        env, "java/util/NoSuchElementException",
        "Cannot invoke removeFirst on an empty deque!");
    return NULL;
  }

  const std::string& str = deque->front();
  jbyteArray elem = rocksdb::JniUtil::stdStringToByteArray(env, str);

  deque->pop_front();

  return elem;
}

/*
 * Class:     org_rocksdb_ByteArrayDeque
 * Method:    removeLast0
 * Signature: (J)[B
 */
jbyteArray JNICALL Java_org_rocksdb_ByteArrayDeque_removeLast0(JNIEnv* env,
                                                               jobject jobj,
                                                               jlong handle) {
  const auto deque = reinterpret_cast<std::deque<std::string> *>(handle);

  if (deque->empty()) {
    rocksdb::ExceptionJni::ThrowNew(
        env, "java/util/NoSuchElementException",
        "Cannot invoke removeLast on an empty deque!");
    return NULL;
  }

  std::string& str(deque->back());
  jbyteArray elem = rocksdb::JniUtil::stdStringToByteArray(env, str);

  deque->pop_back();

  return elem;
}

/*
 * Class:     org_rocksdb_ByteArrayDeque
 * Method:    pollFirst0
 * Signature: (J)[B
 */
jbyteArray JNICALL Java_org_rocksdb_ByteArrayDeque_pollFirst0(JNIEnv* env,
                                                              jobject jobj,
                                                              jlong handle) {
  const auto deque = reinterpret_cast<std::deque<std::string> *>(handle);

  if (deque->empty()) {
    return NULL;
  }

  const std::string& str = deque->front();
  jbyteArray elem = rocksdb::JniUtil::stdStringToByteArray(env, str);

  return elem;
}

/*
 * Class:     org_rocksdb_ByteArrayDeque
 * Method:    pollLast0
 * Signature: (J)[B
 */
jbyteArray JNICALL Java_org_rocksdb_ByteArrayDeque_pollLast0(JNIEnv* env,
                                                             jobject jobj,
                                                             jlong handle) {
  const auto deque = reinterpret_cast<std::deque<std::string> *>(handle);

  if (deque->empty()) {
    return NULL;
  }

  const std::string& str = deque->back();
  jbyteArray elem = rocksdb::JniUtil::stdStringToByteArray(env, str);

  return elem;
}

/*
 * Class:     org_rocksdb_ByteArrayDeque
 * Method:    getFirst0
 * Signature: (J)[B
 */
jbyteArray JNICALL Java_org_rocksdb_ByteArrayDeque_getFirst0(JNIEnv* env,
                                                             jobject jobj,
                                                             jlong handle) {
  const auto deque = reinterpret_cast<std::deque<std::string> *>(handle);

  if (deque->empty()) {
    rocksdb::ExceptionJni::ThrowNew(
        env, "java/util/NoSuchElementException",
        "Cannot invoke getFirst on an empty deque!");
    return NULL;
  }

  const std::string& str = deque->front();
  jbyteArray elem = rocksdb::JniUtil::stdStringToByteArray(env, str);

  return elem;
}

/*
 * Class:     org_rocksdb_ByteArrayDeque
 * Method:    getLast0
 * Signature: (J)[B
 */
jbyteArray JNICALL Java_org_rocksdb_ByteArrayDeque_getLast0(JNIEnv* env,
                                                            jobject jobj,
                                                            jlong handle) {
  const auto deque = reinterpret_cast<std::deque<std::string> *>(handle);

  if (deque->empty()) {
    rocksdb::ExceptionJni::ThrowNew(env, "java/util/NoSuchElementException",
                                    "Cannot invoke getLast on an empty deque!");
    return NULL;
  }

  const std::string& str = deque->back();
  jbyteArray elem = rocksdb::JniUtil::stdStringToByteArray(env, str);

  return elem;
}

/*
 * Class:     org_rocksdb_ByteArrayDeque
 * Method:    getLast0
 * Signature: (J)[B
 */
jbyteArray JNICALL Java_org_rocksdb_ByteArrayDeque_peekFirst0(JNIEnv* env,
                                                              jobject jobj,
                                                              jlong handle) {
  const auto deque = reinterpret_cast<std::deque<std::string> *>(handle);

  if (deque->empty()) {
    return NULL;
  }

  const std::string& str = deque->front();
  jbyteArray elem = rocksdb::JniUtil::stdStringToByteArray(env, str);

  return elem;
}

/*
 * Class:     org_rocksdb_ByteArrayDeque
 * Method:    getLast0
 * Signature: (J)[B
 */
jbyteArray JNICALL Java_org_rocksdb_ByteArrayDeque_peekLast0(JNIEnv* env,
                                                             jobject jobj,
                                                             jlong handle) {
  const auto deque = reinterpret_cast<std::deque<std::string> *>(handle);

  if (deque->empty()) {
    return NULL;
  }

  const std::string& str = deque->back();
  jbyteArray elem = rocksdb::JniUtil::stdStringToByteArray(env, str);

  return elem;
}

/*
 * Class:     org_rocksdb_ByteArrayDeque
 * Method:    removeFirstOccurrence0
 * Signature: (JLjava/lang/Object;)Z
 */
jboolean JNICALL Java_org_rocksdb_ByteArrayDeque_removeFirstOccurrence0(
    JNIEnv* env, jobject jobj, jlong handle, jobject elem) {
  rocksdb::ExceptionJni::ThrowNew(env,
                                  "java/lang/UnsupportedOperationException",
                                  "");
  return false;
}

/*
 * Class:     org_rocksdb_ByteArrayDeque
 * Method:    removeLastOccurrence0
 * Signature: (JLjava/lang/Object;)Z
 */
jboolean JNICALL Java_org_rocksdb_ByteArrayDeque_removeLastOccurrence0(
    JNIEnv* env, jobject jobj, jlong handle, jobject elem) {
  rocksdb::ExceptionJni::ThrowNew(env,
                                  "java/lang/UnsupportedOperationException",
                                  "");
  return false;
}

/*
 * Class:     org_rocksdb_ByteArrayDeque
 * Method:    remove0
 * Signature: (JLjava/lang/Object;)Z
 */
jboolean JNICALL Java_org_rocksdb_ByteArrayDeque_remove0(JNIEnv* env,
                                                         jobject jobj,
                                                         jlong handle,
                                                         jobject elem) {
  rocksdb::ExceptionJni::ThrowNew(env,
                                  "java/lang/UnsupportedOperationException",
                                  "");
  return false;
}

/*
 * Class:     org_rocksdb_ByteArrayDeque
 * Method:    containsAll0
 * Signature: (JLjava/util/Collection;)Z
 */
jboolean JNICALL Java_org_rocksdb_ByteArrayDeque_containsAll0(JNIEnv* env,
                                                              jobject jobj,
                                                              jlong handle,
                                                              jobject elem) {
  rocksdb::ExceptionJni::ThrowNew(env,
                                  "java/lang/UnsupportedOperationException",
                                  "");
  return false;
}

/*
 * Class:     org_rocksdb_ByteArrayDeque
 * Method:    addAll0
 * Signature: (JLjava/util/Collection;)Z
 */
jboolean JNICALL Java_org_rocksdb_ByteArrayDeque_addAll0(JNIEnv* env,
                                                         jobject jobj,
                                                         jlong handle,
                                                         jobject elem) {
  rocksdb::ExceptionJni::ThrowNew(env,
                                  "java/lang/UnsupportedOperationException",
                                  "");
  return false;
}

/*
 * Class:     org_rocksdb_ByteArrayDeque
 * Method:    removeAll0
 * Signature: (JLjava/util/Collection;)Z
 */
jboolean JNICALL Java_org_rocksdb_ByteArrayDeque_removeAll0(JNIEnv* env,
                                                            jobject jobj,
                                                            jlong handle,
                                                            jobject elem) {
  rocksdb::ExceptionJni::ThrowNew(env,
                                  "java/lang/UnsupportedOperationException",
                                  "");
  return false;
}

/*
 * Class:     org_rocksdb_ByteArrayDeque
 * Method:    retainAll0
 * Signature: (JLjava/util/Collection;)Z
 */
jboolean JNICALL Java_org_rocksdb_ByteArrayDeque_retainAll0(JNIEnv* env,
                                                            jobject jobj,
                                                            jlong handle,
                                                            jobject elem) {
  rocksdb::ExceptionJni::ThrowNew(env,
                                  "java/lang/UnsupportedOperationException",
                                  "");
  return false;
}

/*
 * Class:     org_rocksdb_ByteArrayDeque
 * Method:    clear0
 * Signature: (J)V
 */
void JNICALL Java_org_rocksdb_ByteArrayDeque_clear0(JNIEnv* env, jobject jobj,
                                                    jlong handle) {
  const auto deque = reinterpret_cast<std::deque<std::string> *>(handle);

  deque->clear();
}

/*
 * Class:     org_rocksdb_ByteArrayDeque
 * Method:    contains0
 * Signature: (JLjava/lang/Object;)Z
 */
jboolean JNICALL Java_org_rocksdb_ByteArrayDeque_contains0(JNIEnv* env,
                                                           jobject jobj,
                                                           jlong handle,
                                                           jobject elem) {
  rocksdb::ExceptionJni::ThrowNew(env,
                                  "java/lang/UnsupportedOperationException",
                                  "");
  return false;
}

/*
 * Class:     org_rocksdb_ByteArrayDeque
 * Method:    size0
 * Signature: (J)I
 */
jint JNICALL Java_org_rocksdb_ByteArrayDeque_size0(JNIEnv* env, jobject jobj,
                                                   jlong handle) {
  const auto deque = reinterpret_cast<std::deque<std::string> *>(handle);

  return deque->size();
}

/*
 * Class:     org_rocksdb_ByteArrayDeque
 * Method:    isEmpty0
 * Signature: (J)Z
 */
jboolean JNICALL Java_org_rocksdb_ByteArrayDeque_isEmpty0(JNIEnv* env,
                                                          jobject jobj,
                                                          jlong handle) {
  const auto deque = reinterpret_cast<std::deque<std::string> *>(handle);

  return deque->empty();
}

/*
 * Class:     org_rocksdb_ByteArrayDeque
 * Method:    toArray0
 * Signature: (J)[Ljava/lang/Object;
 */
jobjectArray JNICALL Java_org_rocksdb_ByteArrayDeque_toArray0(JNIEnv* env,
                                                              jobject jobj,
                                                              jlong handle) {
  rocksdb::ExceptionJni::ThrowNew(env,
                                  "java/lang/UnsupportedOperationException",
                                  "");
  return NULL;
}

/*
 * Class:     org_rocksdb_ByteArrayDeque
 * Method:    toArray1
 * Signature: (J[Ljava/lang/Object;)[Ljava/lang/Object;
 */
jobjectArray JNICALL Java_org_rocksdb_ByteArrayDeque_toArray1(
    JNIEnv* env, jobject jobj, jlong handle, jobjectArray array) {
  rocksdb::ExceptionJni::ThrowNew(env,
                                  "java/lang/UnsupportedOperationException",
                                  "");
  return NULL;
}

/*
 * Class:     org_rocksdb_ByteArrayDeque
 * Method:    toString0
 * Signature: (JZ)Ljava/lang/String;
 */
jstring JNICALL Java_org_rocksdb_ByteArrayDeque_toString0(JNIEnv* env,
                                                          jobject jobj,
                                                          jlong handle,
                                                          jboolean hex) {
  rocksdb::ExceptionJni::ThrowNew(env,
                                  "java/lang/UnsupportedOperationException",
                                  "");
  return NULL;
}

/*
 * Class:     org_rocksdb_ByteArrayDeque
 * Method:    disposeInternal
 * Signature: (J)V
 */
void JNICALL Java_org_rocksdb_ByteArrayDeque_disposeInternal(JNIEnv* env,
                                                             jobject jobj,
                                                             jlong handle) {
  delete reinterpret_cast<std::deque<std::string> *>(handle);
}

////////////////////////////// Iterator implementation
/*
 * Class:     org_rocksdb_ByteArrayDeque_Iter
 * Method:    itrhasNext0
 * Signature: (JI)Z
 */
jboolean JNICALL Java_org_rocksdb_ByteArrayDeque_00024Iter_itrhasNext0
(JNIEnv* env, jobject jobj, jlong handle, jint idx) {
  const auto deque = reinterpret_cast<std::deque<std::string> *>(handle);
  auto iter = deque->begin() + idx;
  return (iter >= deque->end());
}

/*
 * Class:     org_rocksdb_ByteArrayDeque_Iter
 * Method:    itrNext0
 * Signature: (JI)[B
 */
jbyteArray JNICALL Java_org_rocksdb_ByteArrayDeque_00024Iter_itrNext0
(JNIEnv* env, jobject jobj, jlong handle, jint idx) {
  const auto deque = reinterpret_cast<std::deque<std::string> *>(handle);
  auto iter = deque->begin() + idx;
  if (iter >= deque->end()) {
    rocksdb::ExceptionJni::ThrowNew(
            env, "java/util/NoSuchElementException",
            "Cannot invoke iterator.next() past the last element!");
        return NULL;
  }

  jbyteArray elem = rocksdb::JniUtil::stdStringToByteArray(env, *iter);
  return elem;
}

/*
 * Class:     org_rocksdb_ByteArrayDeque_Iter
 * Method:    itrRemove0
 * Signature: (JI)V
 */
void JNICALL Java_org_rocksdb_ByteArrayDeque_00024Iter_itrRemove0
(JNIEnv* env, jobject jobj, jlong handle, jint idx) {
  const auto deque = reinterpret_cast<std::deque<std::string> *>(handle);
  auto iter = deque->begin() + idx;
  deque->erase(iter);
}


